tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -190,6 +236,361 @@ void BuildDotClusters(Block* block,
+    //get one-side operand shareinfo according to the share_type
+    DotShareInfo share_info;
+    share_info.share_operand = (share_type == LEFT) ? op.lhs() : op.rhs();
+    const auto& dim_numbers = op.dot_dimension_numbers();
+    const auto& lhs_contracting_dims = dim_numbers.getLhsContractingDimensions();
+    const auto& rhs_contracting_dims = dim_numbers.getRhsContractingDimensions();
+    assert(lhs_contracting_dims.size() == 1);
+    share_info.lhs_contracting_dim = lhs_contracting_dims[0];
+    assert(rhs_contracting_dims.size() == 1);
+    share_info.rhs_contracting_dim = rhs_contracting_dims[0];
+    auto& shared_op_list = share_operand_map[std::move(share_info)];
+    shared_op_list.push_back(op);
+  });
+  return true;
+}
+
+
+bool DotShareOperandMergeConverter::applyMerging(DotCluster& cluster, ShareType share_type) {
+  auto& ops = cluster.ops;
+  auto loc = ops.front()->getLoc();
+  auto foremost = ops.front();
+  for (int64_t i = 1; i < ops.size(); i++) {
+    auto& op = ops[i];
+    if (op->isBeforeInBlock(foremost)) {
+      foremost = op;
+    }
+  }
+  // Move all dot ops, and their consumers if necessary, before the original
+  // foremost dot. This makes sure that the newly created ops in this function
+  // dominates their uses.
+  for (auto op : ops) {
+    if (foremost == op) {
+      continue;
+    }
+    op->moveBefore(foremost);
+    ArrangeOperandsInsertPointInBlock(op);
+  }
+  auto foremost_dot = dyn_cast<mhlo::DotGeneralOp>(foremost);
+  // We use the foremost dot to create the builder. Thus we only need to reorder
+  // the operands of some newly created ops, rather users of them.
+  OpBuilder builder(foremost_dot);
+  auto orig_lhs_type = foremost_dot.lhs().getType().dyn_cast<RankedTensorType>();
+  auto orig_rhs_type = foremost_dot.rhs().getType().dyn_cast<RankedTensorType>();
+  auto orig_result_type = foremost_dot.getType().dyn_cast<RankedTensorType>();
+  SmallVector<Value> lhs_operands;
+  SmallVector<Value> rhs_operands;
+  auto m_dim = getMNDim(foremost_dot).first;
+  auto n_dim = getMNDim(foremost_dot).second;
+  int64_t concat_dim = (share_type == LEFT) ? n_dim : m_dim;              
+  bool is_dynamic_shape = false;
+  int64_t concat_dim_sum = 0;
+  Value lhs;
+  Value rhs;
+  if(share_type == LEFT){
+    for (auto op : ops) {
+      mhlo::DotGeneralOp dot = dyn_cast<mhlo::DotGeneralOp>(op);
+      auto rhs_type = dot.rhs().getType().dyn_cast<RankedTensorType>();
+      auto concat_dim_size = rhs_type.getDimSize(concat_dim);
+      if (concat_dim_size == -1) {
+        is_dynamic_shape = true;
+      } else {
+        concat_dim_sum += concat_dim_size;
+
+      }
+      rhs_operands.push_back(dot.rhs());
+    }
+    
+    //concat rhs
+    auto rhs_rank = orig_rhs_type.getRank();
+    SmallVector<int64_t, 4> rhs_shapes(rhs_rank, ShapedType::kDynamicSize);
+    for (int64_t i = 0; i < rhs_rank; i++) {
+      if(i == concat_dim) {
+        if (is_dynamic_shape) {
+          rhs_shapes[i] = -1;
```

Comment:
Do not need to update as the value is already `ShapedType::kDynamicSize` when building the vector.